scapy/contrib/rpl_metrics.py

```diff
@@ -0,0 +1,291 @@
+# This file is part of Scapy.
+# See http://www.secdev.org/projects/scapy for more information.
+#
+# Scapy is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# Scapy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Scapy.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Copyright (C) 2020 Rahul Jadhav <nyrahul@gmail.com>
+
+# RFC 6551
+# scapy.contrib.description = Routing Metrics used for Path Calc in LLNs
+# scapy.contrib.status = loads
+
+"""
+RFC 6551 - Routing Metrics Used for Path Calculation in LLNs
+
++----------------------------+
+| Metrics & Constraint Types |
++----------------------------+
+| DAGMC Option               |
++----------------------------+
+| RPL-DIO                    |
++----------------------------+
+"""
+
+import struct
+from scapy.compat import orb
+from scapy.packet import Packet, bind_layers
+from scapy.fields import ByteEnumField, ByteField, ShortField, BitField, \
+    BitEnumField, FieldLenField, StrLenField, IntField
+from scapy.layers.inet6 import _PhantomAutoPadField, _OptionsField
+from scapy.contrib.rpl import RPLOPTSSTR, _RPLGuessMsgType
+
+
+class _DAGMetricContainer(Packet):
+    name = 'Dummy DAG Metric container'
+
+    def post_build(self, p, pay):
+        p += pay
+        tmp_len = self.len
+        if self.len is None:
+            tmp_len = len(p) - 2
+        p = p[:1] + struct.pack("B", tmp_len) + p[2:]
+        return p
+
+
+DAGMC_OBJTYPE = {1: "Node State and Attributes",
+                 2: "Node Energy",
+                 3: "Hop Count",
+                 4: "Link Throughput",
+                 5: "Link Latency",
+                 6: "Link Quality Level",
+                 7: "Link ETX",
+                 8: "Link Color"}
+
+
+class DAGMCObjUnknown(Packet):
+    """
+    Dummy unknown metric/constraint
+    """
+    name = 'Unknown DAGMC Object Option'
+    fields_desc = [ByteEnumField("otype", 3, DAGMC_OBJTYPE),
+                   FieldLenField("olen", None, length_of="odata", fmt="B"),
+                   StrLenField("odata", "",
+                               length_from=lambda pkt: pkt.olen)]
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *_, **kargs):
```

Comment:
It seems that this is never reached by your tests (https://codecov.io/gh/secdev/scapy/pull/2663/diff?src=pr&el=tree#diff-c2NhcHkvY29udHJpYi9ycGxfbWV0cmljcy5weQ==). Could you find a way to test this?

Reply:
Vastly improved UT in the update.